modules/navigation/navigation_geometry_parser_3d.h

```diff
@@ -0,0 +1,70 @@
+/*************************************************************************/
+/*  navigation_geometry_parser_3d.h                                      */
+/*************************************************************************/
+/*                       This file is part of:                           */
+/*                           GODOT ENGINE                                */
+/*                      https://godotengine.org                          */
+/*************************************************************************/
+/* Copyright (c) 2007-2022 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2022 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef NAVIGATION_GEOMETRY_PARSER_3D_H
+#define NAVIGATION_GEOMETRY_PARSER_3D_H
+
+#include "core/templates/vector.h"
+#include "scene/3d/node_3d.h"
+#include "scene/resources/navigation_mesh.h"
+
+class NavigationGeometryParser3D : public RefCounted {
+	GDCLASS(NavigationGeometryParser3D, RefCounted);
+	friend class NavigationMeshGenerator;
+
+protected:
+	static void _bind_methods();
+
+	GDVIRTUAL1RC(bool, _parses_node, Node *);
+
+	GDVIRTUAL2C(_parse_geometry, Node *, Ref<NavigationMesh>);
```

Comment:
Right now, parsers need to check the `parsed_geometry_type ` to know if they should be looking for collision shapes or visual meshes.

What if instead, we have two methods here:
- `_parse_geometry` (would parse visual geometry)
- `_parse_collision` (would parse only collision shapes)

Then, if a parser supports only visual meshes, it would implement `_parse_geometry`.  If it only supports collision shapes, it would implement `_parse_collision`.  If it supports both, it implements both.

This should simplify the implementation of parsers and might remove the need to pass in the target `NavigationMesh` at all.  The base parser would be in charge of calling the correct method based on the `parsed_geometry_type` field.

Reply:
I don't think we can stop passing the NavigationMesh ref as long as this object holds the user set bake properties. Users might want to change the added source geometry entirely depending on some of those values e.g. snap to cell value or simplify depending on slope value.

While most of the code that checks for parsed_geometry_type was a copy&paste from existing parsing I am not sure adding this function distinction between collision and everything else has any real benefit. The source geometry parser does not care what kind of object adds source geometry, could be just a basic node that does it, so this distinction would be entirely artifical. Having a single function for everything keeps things simple and allows addon / module devs to do whatever they want. When you have specific functions and someone has a shared node class that wants to add both visuals and collisions in a very specific order you are already creating obstacles for them by splitting functions with a fixed order for nothing.